### ISSUE TITLE
Integer Overflow (?) on implementation pp. 127, in section 5.3.3 Hash Codes for Arrays and Strings, C++ edition

### ISSUE THREAD

**Issue by EigenVecta** [patmorin/ods#99]
I suspect the line
```
long p = (1L<<32)-5; // prime: 2ˆ32 - 5
```
causes overflow. ``long`` on my machine is 32-bit (``printf("%ld\n", LONG_MAX);`` prints 2147483647), and compiler would give a warning ``-Wshift-count-overflow`` on that line. Indeed ``long`` type cannot hold that prime number.

I suspect it could be corrected as
```
unsigned long p = (1L<<32) - 5;
```
with ``printf("%lu\n", p);`` prints 4294967291, which is desired.

However the ``-Wshift-count-overflow`` warning: left shift count >= width of type still persists.

With
```
unsigned long p = (~0L) - 4;
```
(*)
the warning vanishes.

In summary, I think that line should be (*).
